smart-contracts/contracts/Unlock.sol

```diff
@@ -488,6 +506,126 @@ contract Unlock is UnlockInitializable, UnlockOwnable {
     }
   }
 
+  function getBalance(address token) internal view returns (uint) {
+    return token == address(0) ?
+      address(this).balance 
+      :
+      IMintableERC20(token).balanceOf(address(this));
+  }
+
+  // set permit2 address
+  function setPermit2(address _permit2) public onlyOwner {
+     permit2 = _permit2;
+  }
+
+  /**
+   * Swap tokens and call a function in a lock contract
+   * @notice If the actual amount spent is less than the specified maximum amount, the remaining tokens wll 
+   * be held by the Unlock contract
+   */
+  function swapAndCall(
+    address lock,
+    address srcToken,
+    uint amountInMax,
+    address swapRouter,
+    bytes memory swapCalldata,
+    bytes memory callData
+  ) public payable returns(bytes memory) {
+    // check if lock exists
+    if(!locks[lock].deployed) {
+      revert LockDoesntExist(lock);
+    }
+
+    // get lock pricing 
+    address destToken = IPublicLock(lock).tokenAddress();
+
+    // get price in destToken from lock
+    uint keyPrice = IPublicLock(lock).keyPrice();
+
+    // get balances of Unlock before
+    // if payments in ETH, substract the value sent by user to get actual balance
+    uint balanceTokenDestBefore = destToken == address(0) ? 
+      getBalance(destToken) - msg.value 
+            :
+      getBalance(destToken);
+
+    uint balanceTokenSrcBefore = 
+        srcToken == address(0) ? 
+          getBalance(srcToken) - msg.value 
+          :
+          getBalance(srcToken);
+
+    if(srcToken != address(0)) {
+      // Transfer the specified amount of src ERC20 to this contract
+      TransferHelper.safeTransferFrom(srcToken, msg.sender, address(this), amountInMax);
+
+      // Approve the router to spend src ERC20
+      TransferHelper.safeApprove(srcToken, swapRouter, amountInMax);
+
+      // approve PERMIT2 to mainpulate the token
+      IERC20(srcToken).approve(permit2, amountInMax);
+    }
+
+    // issue PERMIT2 Allowance
+    IPermit2(permit2).approve(
+      srcToken,
+      swapRouter,
+      uint160(amountInMax),
+      uint48(block.timestamp + 3600) // expiration
```

Comment:
Any reason to make this 1 hour? Technically this should happen inside this block, right? If so then let's reduce to say 1 minute?

Reply:
no specific reason, yes lets make 1min.